probe-rs/src/architecture/arm/memory/adi_v5_memory_interface.rs

```diff
@@ -693,71 +741,89 @@ where
+
+        Ok(())
+    }
+
+    /// Write a block of 16 bit words at `address`.
+    ///
+    /// The number of words written is `data.len()`.
+    /// The address where the write should be performed at has to be word aligned.
+    /// Returns `ArmError::MemoryNotAligned` if this does not hold true.
+    pub fn write_16(
+        &mut self,
+        access_port: MemoryAp,
+        mut address: u64,
+        mut data: &[u16],
+    ) -> Result<(), ArmError> {
+        if self.ap_information.supports_only_32bit_data_size {
+            return Err(ArmError::UnsupportedTransferWidth(8));
+        }
+        if (address % 2) != 0 {
+            return Err(ArmError::alignment_error(address, 2));
+        }
+        if data.is_empty() {
+            return Ok(());
+        }
 
         tracing::debug!(
-            "Write first block with len {} at address {:#08x}",
-            first_chunk_size_bytes,
+            "Write block with total size {} bytes to address {:#08x}",
+            data.len() * 2,
             address
         );
 
-        let first_chunk_size_transfer_unit = first_chunk_size_bytes / 4;
-
-        self.write_ap_register_repeated(
-            access_port,
-            DRW { data: 0 },
-            &data[data_offset..first_chunk_size_transfer_unit],
-        )?;
-
-        remaining_data_len -= first_chunk_size_transfer_unit;
-        let mut address = address
-            .checked_add((first_chunk_size_transfer_unit * 4) as u64)
-            .ok_or(ArmError::OutOfBounds)?;
-        data_offset += first_chunk_size_transfer_unit;
-
-        while remaining_data_len > 0 {
-            // the autoincrement is limited to the 10 lowest bits so we need to write the address
-            // every time it overflows
-            self.write_tar_register(access_port, address)?;
+        let csw = self.build_csw_register(DataSize::U16);
+        self.write_csw_register(access_port, csw)?;
 
-            let next_chunk_size_bytes = std::cmp::min(max_chunk_size_bytes, remaining_data_len * 4);
+        while !data.is_empty() {
+            let chunk_size = data.len().min(autoincr_max_bytes(address) / 2);
 
             tracing::debug!(
                 "Writing chunk with len {} at address {:#08x}",
-                next_chunk_size_bytes,
+                chunk_size,
                 address
             );
 
-            let next_chunk_size_transfer_unit = next_chunk_size_bytes / 4;
+            // The required shifting logic here is described in C2.2.6 Byte lanes of the ADI v5.2 specification.
+            // All bytes are transfered in their lane, so when we do an access at an address that is not divisible by 4,
+            // we have to shift the word (one or two bytes) to it's correct position.
+            let values = data[..chunk_size]
+                .iter()
+                .enumerate()
+                .map(|(i, v)| (*v as u32) << (((address as usize + i * 2) % 4) * 8))
+                .collect::<Vec<_>>();
```

Comment:
This seems quite complex, could you add some tests for this?

Reply:
Added tests for 16bit accesses.